### DISCUSSION TITLE
zero-length arrays in io_uring.h

### DISCUSSION THREAD

**Discussion by ghost** [axboe/liburing#1034] [Q&A] [answered]
C doesn't have generics, but zero-length arrays make the entire structure unable to be copied. Is this intentional? I mean, would it be better if we had io_uring_cqe_16, io_uring_cqe_32 instead of the existing io_uring_cqe?

**Comment by DylanZA** [comment, marked as answer]
the size is runtime dependent (based on IORING_SETUP_CQE32), so you cannot have it typed safely

**Reply by ghost** [reply]
This problem does not exist because a zero-length array also cannot retrieve its length when IORING_SETUP_CQE32 is not available, as it does not store length.

**Reply by ghost** [reply]
If generics were used, it would be like this: `io_uring<S, C>`, where `S` can be either `SQE64` or `SQE128`, and `C` can be either `CQE16` or `CQE32`.
